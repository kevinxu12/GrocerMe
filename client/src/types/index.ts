/**
 * @file Top-level entry type for the types directory
 * @author Kevin Xu
 */
import { RootState } from './RootState';
export type setMessageSnackbarType = (string) => void;
export type { RootState };
