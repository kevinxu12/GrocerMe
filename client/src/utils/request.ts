/**
 * @file Helpers related to making http requests from the frontend
 * @author Kevin Xu
 */
/* eslint-disable jsdoc/require-returns-type */
import axios, { AxiosResponse } from 'axios';
import generateServerUrl from './url';

/**
 * Parses a raw axios response to fetch data stored within our custom SuccessResponse
 *
 * @param {AxiosResponse} response raw Axios Response
 * @returns Custom data stored in the SuccessResponse, else Null
 */
export function parseAxiosSuccessResponse<T>(
  response: AxiosResponse,
): T | null {
  if (response.status === 204 || response.status === 205) {
    return null;
  }
  const responseData = response.data;
  // custom data will be stored in here
  if ('data' in responseData) {
    return responseData.data;
  }
  return null;
}

/**
 * Checks if a network request came back fine, and throws an error if not
 *
 * @param  {object} response   A response from a network request
 * @returns {object|undefined} Returns either the response, or throws an error
 */
export function checkStatus(response) {
  if (response.status >= 200 && response.status < 300) {
    return response;
  }

  const error = new Error(response.statusText);
  throw error;
}

/**
 * Timeout for some length
 *
 * @param {*} ms time of delay
 * @returns {*} Promise of a timeout
 */
export function timeout(ms) {
  return new Promise(resolve => setTimeout(resolve, ms));
}

/**
 * An Api client wrapper around axios. Use this instead of calling axios directly
 */
export class Api {
  /**
   * Wrapper around axios.get
   *
   * @param {string} url the raw url string
   * @returns {AxiosResponse} axiosResponse
   */
  public static async get(url: string) {
    return await axios.get(generateServerUrl(url), { withCredentials: true });
  }

  /**
   * Wrapper around axios.post
   *
   * TO DO, we should probably type the request body, instead of using ANY in the future.
   *
   * @param {string} url raw url string
   * @param {any} body Post object
   * @returns {AxiosResponse} axiosResponse
   */
  public static async post(url: string, body: any = {}) {
    return await axios.post(generateServerUrl(url), body, {
      withCredentials: true,
    });
  }
}
