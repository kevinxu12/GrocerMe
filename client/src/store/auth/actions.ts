/**
 * @file Auth Actions
 * @author Kevin Xu
 */
import { RoleCode } from 'types/rest';
import { ChangeAuthAction, LogoutAction } from './../../types/actions';
import { CHANGE_AUTH, LOGOUT } from './constants';
/**
 * Changes the logged-in user in the reducer
 *
 * @param  {object} props props
 * @param {string} props.username The new email of the logged-in user
 * @param {RoleCode[]} props.roles Te new roles of the logged-in user
 * @returns {object} An action object with a type of CHANGE_AUTH
 */
export const changeAuth = (props: {
  username: string;
  roles?: RoleCode[];
}): ChangeAuthAction => ({
  type: CHANGE_AUTH,
  payload: {
    ...props,
  },
});

/**
 * Logs a user out
 *
 * @returns {object} An action object with type LOGOUT
 */
export const logout = (): LogoutAction => {
  return {
    type: LOGOUT,
    payload: {},
  };
};

export default changeAuth;
