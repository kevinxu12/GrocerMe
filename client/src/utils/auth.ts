/**
 * @file Generate authentication related helpers
 * @author Kevin Xu
 */

import { Role, RoleCode } from 'types/rest';
import { Constants } from './constants';

export const NO_ROLE_PATHS = [
  Constants.HOME_URL,
  Constants.TEST_SOCKET_URL,
  Constants.ABOUT,
  Constants.CONTACT_US,
];
const ALL_ROLE_PATHS = [Constants.SUPPLIER_HOME, Constants.USER_HOME];
const ADMIN_PATHS = [Constants.ADMIN_HOME];
const CONSUMER_PATHS = [];
const SUPPLIER_PATHS = [];
const role_path_object = {
  ADMIN: ADMIN_PATHS,
  CONSUMER: CONSUMER_PATHS,
  SUPPLIER: SUPPLIER_PATHS,
};

/**
 * Check if a given set of roles is authenticated for a path
 *
 * @param {string} path the path we want to check authentication for
 * @param {Role[]} roles the roles a user is permissioned to
 * @returns {boolean} whether the path is authenticated depending on the roles
 */
export const isPathAuthenticated = (path: string, roles: Role[]): boolean => {
  const role_codes = roles
    .filter((role: Role) => role.status)
    .map((role: Role) => role.code);
  // Admin, Supplier
  let all_paths = role_codes.map(code => role_path_object[code]);
  if (roles.length > 0) {
    all_paths = all_paths.concat(ALL_ROLE_PATHS);
  }
  const all_available_paths: string[] = NO_ROLE_PATHS.concat(
    [].concat(...all_paths),
  );
  return all_available_paths.includes(path);
};

/**
 * Check if a roleCode is contained within a list of roles
 *
 * @param {RoleCode} roleCode RoleCode that may or may not be included
 * @param {Role[]} roles roles to check within
 * @returns {boolean} true if included, false otherwise
 */
export const isRoleCodeIncluded = (
  roleCode: RoleCode,
  roles: Role[] | null,
): boolean => {
  if (!roles) {
    return false;
  }
  const matching_role = roles.filter((role: Role) => role.code === roleCode);
  return matching_role.length > 0;
};
