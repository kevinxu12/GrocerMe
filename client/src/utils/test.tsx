/**
 * @file Helper related to unit testing with jest
 */

import React, { ReactElement } from 'react';
import configureStore from 'redux-mock-store';
import { BrowserRouter as Router } from 'react-router-dom';
import { render, RenderResult } from '@testing-library/react';
import { Provider } from 'react-redux';
import { Store } from '@reduxjs/toolkit';
import { RoleCode } from 'types/rest';
import { Constants } from './constants';
import { AuthState } from 'types/RootState';
/* eslint-disable jsdoc/no-undefined-types */
/**
 * Generates a mock state for our auth reducer
 *
 * @param {RoleCode[]} roles the list of RoleCodes our mock logged in user has
 * @param {string} username the username of our mocked logged in user
 * @returns {AuthState} the mock auth state
 */
export const generateAuthState = (
  roles: RoleCode[],
  username: string = Constants.TEST_USERNAME,
): AuthState => {
  return {
    username: Constants.TEST_USERNAME,
    roles: roles,
  };
};

/**
 * A class that when instantiated, creates a mock provider and store for the entire testing class.
 * All subsequent components rendered with the helper methods access this store
 * Most wrappers will use <T> = <RootState>
 */
export class MockStoreWrapper<T> {
  mockStore: any; // not sure what's the typing here
  state: T; // the store state

  /**
   * Initialize the Mock Store Wrapper
   *
   * @param {T} state The initial state for the store
   */
  constructor(state: T) {
    this.mockStore = configureStore();
    this.state = state;
  }

  /**
   * Create a mock store with an explicit state, default to the state instance variable otherwise
   *
   * @param {T} state an explicit state to keep track of
   * @returns {Store} a mock store that holds the desired state
   */
  public generateStore(state: T = this.state): Store<T> {
    return this.mockStore(state);
  }

  /**
   * Wraps a component with a Provider and Router
   *
   * @param {React.ReactElement} component Component to be wrapped
   * @param {T} state Optional state to create store with
   * @returns {RenderResult} testing library render
   */
  public renderComponentWithRouterProvider(
    component: React.ReactElement,
    state: T = this.state,
  ): RenderResult {
    const store = this.generateStore(state);
    const jsx = (
      <Provider store={store}>
        <Router>{component}</Router>
      </Provider>
    );
    return render(jsx);
  }

  /**
   * Wraps a component with a Provider and Router
   *
   * @param {React.ReactElement} component Component to be wrapped
   * @param {T} state Optional state to create store with
   * @returns {ReactElement} testing library render
   */
  public genComponentWithRouterProvider(
    component: React.ReactElement,
    state: T = this.state,
  ): React.ReactElement {
    const store = this.generateStore(state);
    const jsx = (
      <Provider store={store}>
        <Router>{component}</Router>
      </Provider>
    );
    return jsx;
  }

  /**
   * Wraps a component with a Provider
   *
   * @param {React.ReactElement} component Component to be wrapped
   * @param {T} state Optional state to create store with
   * @returns {RenderResult} testing library render
   */
  public renderComponentWithProvider(
    component: React.ReactElement,
    state: T = this.state,
  ): RenderResult {
    const store = this.generateStore(state);
    return render(<Provider store={store}>{component}</Provider>);
  }
}

/**
 * Test Component for class
 *
 * @returns {React.ReactElement} Test Component
 */
export const TestComponent = (): React.ReactElement => {
  return <div> Test </div>;
};
