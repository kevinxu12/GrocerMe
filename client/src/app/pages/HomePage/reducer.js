/*
 * HomeReducer
 *
 * The reducer takes care of our data. Using actions, we can
 * update our application state. To add a new action,
 * add it to the switch statement in the reducer function
 *
 */

import produce from 'immer';
import { Constants } from 'utils/constants';
import { CHANGE_USERNAME } from './constants';

// The initial state of the App
// const storagedata = JSON.parse(localStorage.getItem("data"));
export const initialState = {
  username: Constants.DEFAULT_STRING,
};

/* eslint-disable default-case, no-param-reassign */
const homeReducer = (state = initialState, action) =>
  produce(state, draft => {
    switch (action.type) {
      case CHANGE_USERNAME:
        // Delete prefixed '@' from the github username
        draft.username = action.username;
        break;
    }
  });

export default homeReducer;
