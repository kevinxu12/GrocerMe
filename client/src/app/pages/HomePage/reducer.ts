/*
 * HomeReducer
 *
 * The reducer takes care of our data. Using actions, we can
 * update our application state. To add a new action,
 * add it to the switch statement in the reducer function
 *
 */

import produce from 'immer';
import { HomeAction } from 'types/actions';
import { HomeState } from 'types/RootState';
import { Constants } from 'utils/constants';
import { CHANGE_USERNAME } from './constants';

// The initial state of the App
export const initialState: HomeState = {
  username: Constants.DEFAULT_STRING,
};

/* eslint-disable default-case, no-param-reassign */
const homeReducer = (
  state: HomeState = initialState,
  action: HomeAction,
): HomeState => {
  return produce(state, draft => {
    switch (action.type) {
      case CHANGE_USERNAME:
        draft.username = action.payload.username;
        break;
    }
  });
};

export default homeReducer;
