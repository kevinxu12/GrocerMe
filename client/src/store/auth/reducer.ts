/**
 * @file HomeReducer
 * @author Kevin Xu
 *
 * The reducer takes care of our data. Using actions, we can
 * update our application state. To add a new action,
 * add it to the switch statement in the reducer function
 */

import produce from 'immer';
import { AuthAction, ChangeUsernameAction } from './../../types/actions';
import { AuthState } from './../../types/RootState';
import { Constants } from './../../utils/constants';
import { CHANGE_USERNAME, LOGOUT } from './constants';

// The initial state of the App
export const initialState: AuthState = {
  username: Constants.DEFAULT_STRING,
};

/* eslint-disable default-case, no-param-reassign */
/**
 * @param {object} state Default state of the reducer
 * @param {object} action Action the reducer listens to
 * @returns {object} the new state of the reducer
 */
const authReducer = (
  state: AuthState = initialState,
  action: AuthAction,
): AuthState => {
  return produce(state, draft => {
    switch (action.type) {
      case CHANGE_USERNAME:
        if (!action.payload) {
          break;
        }
        draft.username = (action as ChangeUsernameAction).payload.username;
        break;
      case LOGOUT:
        return initialState;
    }
  });
};

export default authReducer;
