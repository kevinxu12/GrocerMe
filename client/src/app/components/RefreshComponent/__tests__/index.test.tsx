
import * as React from 'react';
import { initialState } from 'store/auth/reducer';
import { MockStoreWrapper, TestComponent } from 'utils/test';
import { RootState } from 'types';
import { RefreshComponent } from '..';
import { parseAxiosSuccessResponse } from 'utils/request';
import axios from 'axios';
import { render, waitFor } from '@testing-library/react';

jest.mock('axios');
jest.mock('utils/request', () => ({
    __esModule: true,
    parseAxiosSuccessResponse: jest.fn(),
  }))
jest.mock('context/SocketContext', () => ({
    __esModule: true,
    useSocket: () => ({
        emit: jest.fn()
    })
  }))
const changeAuthMock = jest.fn();
const logoutMock = jest.fn();

describe('<RefreshComponent  />', () => {
    const mockStoreWrapper = new MockStoreWrapper<RootState>({auth: initialState});
    it('Logged in user should see state update', async () => {
      (axios.get as jest.Mock).mockImplementation(() => null);
      (parseAxiosSuccessResponse as jest.Mock).mockImplementation(() => ({username: 'hi', roles: []}));
      const pr = render(<RefreshComponent username={'test'} roles={[]} logout={logoutMock} changeAuth={changeAuthMock}><TestComponent/></RefreshComponent>);
      expect(pr.container.firstChild).toMatchSnapshot();
      expect(axios.get).toHaveBeenCalledTimes(1);
      await waitFor(() => expect(parseAxiosSuccessResponse).toHaveBeenCalledTimes(1));
      await waitFor(() => expect(changeAuthMock).toHaveBeenCalledTimes(1));
      await waitFor(() => expect(logoutMock).toHaveBeenCalledTimes(0));
    });
    it('No Logged in user should return nearly', async () => {
      (axios.get as jest.Mock).mockImplementation(() => null);
      (parseAxiosSuccessResponse as jest.Mock).mockImplementation(() => null);
      const pr = mockStoreWrapper.renderComponentWithRouterProvider(<RefreshComponent username={'test'} roles={[]} logout={logoutMock} changeAuth={changeAuthMock}><TestComponent/></RefreshComponent>);
      expect(pr.container.firstChild).toMatchSnapshot();
      expect(axios.get).toHaveBeenCalledTimes(1);
      await waitFor(() => expect(parseAxiosSuccessResponse).toHaveBeenCalledTimes(1));
      await waitFor(() => expect(logoutMock).toHaveBeenCalledTimes(1));
      await waitFor(() => expect(changeAuthMock).toHaveBeenCalledTimes(0));
    });
  });