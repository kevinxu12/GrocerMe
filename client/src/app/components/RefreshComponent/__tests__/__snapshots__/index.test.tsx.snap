// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RefreshComponent  /> Logged in user should see state update 1`] = `
.c0 {
  height: 100%;
  width: 100%;
  margin-left: 20%;
  margin-top: 4rem;
}

<div
  class="c0"
>
  <div>
     Test 
  </div>
</div>
`;

exports[`<RefreshComponent  /> No Logged in user should return nearly 1`] = `
.c0 {
  height: 100%;
  width: 100%;
  margin-left: 20%;
  margin-top: 4rem;
}

<div
  class="c0"
>
  <div>
     Test 
  </div>
</div>
`;
